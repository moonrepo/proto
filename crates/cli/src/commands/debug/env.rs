use crate::components::is_path_like;
use crate::session::ProtoSession;
use clap::Args;
use iocraft::prelude::*;
use proto_core::layout::Store;
use proto_pdk_api::{HostArch, HostOS};
use serde::Serialize;
use starbase::AppResult;
use starbase_console::ui::*;
use starbase_utils::json;
use std::collections::BTreeMap;
use std::env;
use std::path::PathBuf;

#[derive(Serialize)]
struct EnvironmentInfo {
    arch: HostArch,
    configs: Vec<PathBuf>,
    os: HostOS,
    proto_version: String,
    vars: BTreeMap<String, String>,
    virtual_paths: BTreeMap<PathBuf, PathBuf>,
}

#[derive(Serialize)]
struct DebugEnvResult<'a> {
    store: &'a Store,
    env: EnvironmentInfo,
}

#[derive(Args, Clone, Debug)]
pub struct DebugEnvArgs {
    #[arg(long, help = "Print the data in JSON format")]
    json: bool,
}

#[tracing::instrument(skip_all)]
pub async fn env(session: ProtoSession, args: DebugEnvArgs) -> AppResult {
    let env = &session.env;
    let manager = env.load_config_manager()?;

    let environment = EnvironmentInfo {
        arch: HostArch::from_env(),
        configs: manager
            .files
            .iter()
            .filter_map(|file| {
                if file.exists {
                    Some(file.path.to_path_buf())
                } else {
                    None
                }
            })
            .collect(),
        os: HostOS::from_env(),
        proto_version: env!("CARGO_PKG_VERSION").into(),
        vars: env::vars()
            .filter_map(|(k, v)| {
                if k.starts_with("PROTO_") {
                    Some((k, v))
                } else {
                    None
                }
            })
            .collect(),
        virtual_paths: env.get_virtual_paths(),
    };

    if args.json {
        let result = DebugEnvResult {
            store: &env.store,
            env: environment,
        };

        session
            .console
            .out
            .write_line(json::format(&result, true)?)?;

        return Ok(None);
    }

    let store_paths = vec![
        ("Root", &env.store.dir),
        ("Bins", &env.store.bin_dir),
        ("Shims", &env.store.shims_dir),
        ("Plugins", &env.store.plugins_dir),
        ("Tools", &env.store.inventory_dir),
        ("Temp", &env.store.temp_dir),
    ];

    session.console.render(element! {
        Container {
            Section(title: "Store") {
                #(store_paths.into_iter().map(|(name, path)| {
                    element! {
                        Entry(
                            name,
                            value: element! {
                                StyledText(content: path.to_string_lossy(), style: Style::Path)
                            }.into_any(),
                        )
                    }
                }))
            }
            Section(title: "Environment") {
                Entry(
                    name: "Proto version",
                    content: environment.proto_version,
                )
                Entry(
                    name: "Operating system",
                    content: environment.os.to_string(),
                )
                Entry(
                    name: "Architecture",
                    content: environment.arch.to_string(),
                )
                Entry(
                    name: "Config sources",
                    no_children: environment.configs.is_empty(),
                ) {
                    List {
                        #(environment.configs.into_iter().map(|file| {
                            element! {
                                ListItem {
                                    StyledText(
                                        content: file.to_string_lossy(),
                                        style: Style::Path,
                                    )
                                }
                            }
                        }))
                    }
                }
                Entry(
                    name: "Virtual paths",
                    no_children: environment.virtual_paths.is_empty(),
                ) {
                    Map {
                        #(environment.virtual_paths.into_iter().map(|(host, guest)| {
                            let name = element! {
                                StyledText(
                                    content: guest.to_string_lossy(),
                                    style: Style::File
                                )
                            }.into_any();

                            let value = element! {
                                StyledText(
                                    content: host.to_string_lossy(),
                                    style: Style::Path
                                )
                            }.into_any();

                            element! {
                                MapItem(name, value)
                            }
                        }))
                    }
                }
                Entry(
                    name: "Environment variables",
                    no_children: environment.vars.is_empty(),
                ) {
                    Map {
                        #(environment.vars.into_iter().map(|(name, value)| {
                            let name = element! {
                                StyledText(
                                    content: name,
                                    style: Style::Property
                                )
                            }.into_any();

                            let value = element! {
                                StyledText(
                                    content: &value,
                                    style: if is_path_like(&value) {
                                        Style::Path
                                    } else {
                                        Style::MutedLight
                                    }
                                )
                            }.into_any();

                            element! {
                                MapItem(name, value)
                            }
                        }))
                    }
                }
            }
        }
    })?;

    Ok(None)
}
