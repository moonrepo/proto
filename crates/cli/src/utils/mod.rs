pub mod install_graph;
pub mod tool_record;
