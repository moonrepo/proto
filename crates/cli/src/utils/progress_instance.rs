use starbase_console::ConsoleError;
use starbase_console::ui::ProgressReporter;
use std::ops::Deref;
use std::sync::Arc;
use tokio::task::JoinHandle;

pub struct ProgressInstance {
    pub handle: JoinHandle<Result<(), ConsoleError>>,
    pub reporter: Arc<ProgressReporter>,
}

impl ProgressInstance {
    pub async fn stop(self) -> Result<(), ConsoleError> {
        self.reporter.exit();

        if let Ok(result) = self.handle.await {
            result?;
        }

        Ok(())
    }
}

impl Deref for ProgressInstance {
    type Target = ProgressReporter;

    fn deref(&self) -> &Self::Target {
        &self.reporter
    }
}
