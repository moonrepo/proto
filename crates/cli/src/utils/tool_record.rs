use core::ops::{Deref, DerefMut};
use proto_core::{ProtoConfig, ProtoToolConfig, Tool, UnresolvedVersionSpec, VersionSpec};
use std::collections::BTreeMap;

pub struct ToolRecord {
    pub tool: Tool,
    pub config: ProtoToolConfig,
    pub installed_versions: Vec<VersionSpec>,
    pub local_aliases: BTreeMap<String, UnresolvedVersionSpec>,
    pub remote_aliases: BTreeMap<String, UnresolvedVersionSpec>,
    pub remote_versions: Vec<VersionSpec>,
}

impl ToolRecord {
    pub fn new(tool: Tool) -> Self {
        let mut versions = tool
            .inventory
            .manifest
            .installed_versions
            .iter()
            .cloned()
            .collect::<Vec<_>>();
        versions.sort();

        Self {
            tool,
            config: ProtoToolConfig::default(),
            local_aliases: BTreeMap::default(),
            remote_aliases: BTreeMap::default(),
            installed_versions: versions,
            remote_versions: vec![],
        }
    }

    pub fn inherit_from_local(&mut self, config: &ProtoConfig) {
        if let Some(tool_config) = config.tools.get(&self.id).map(|c| c.to_owned()) {
            self.local_aliases.extend(tool_config.aliases.clone());
            self.config = tool_config;
        }
    }

    pub async fn inherit_from_remote(&mut self) -> miette::Result<()> {
        let version_resolver = self
            .tool
            .load_version_resolver(&UnresolvedVersionSpec::default())
            .await?;

        self.remote_aliases.extend(version_resolver.aliases);
        self.remote_versions.extend(version_resolver.versions);
        self.remote_versions.sort();

        Ok(())
    }
}

impl Deref for ToolRecord {
    type Target = Tool;

    fn deref(&self) -> &Self::Target {
        &self.tool
    }
}

impl DerefMut for ToolRecord {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.tool
    }
}
