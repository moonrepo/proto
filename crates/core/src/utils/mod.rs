pub mod archive;
pub mod git;
pub mod process;
