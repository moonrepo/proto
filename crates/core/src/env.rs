use crate::config::{ConfigMode, PROTO_CONFIG_NAME, PinLocation, ProtoConfig};
use crate::config_error::ProtoConfigError;
use crate::env_error::ProtoEnvError;
use crate::file_manager::{ProtoConfigFile, ProtoDirEntry, ProtoFileManager};
use crate::helpers::is_offline;
use crate::layout::Store;
use once_cell::sync::OnceCell;
use starbase_console::{Console, EmptyReporter};
use starbase_utils::dirs::home_dir;
use starbase_utils::env::path_var;
use std::collections::BTreeMap;
use std::env;
use std::fmt;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Duration;
use tracing::debug;
use warpgate::PluginLoader;

pub type ProtoConsole = Console<EmptyReporter>;

#[derive(Clone, Default)]
pub struct ProtoEnvironment {
    pub config_mode: ConfigMode,
    pub env_mode: Option<String>,
    pub home_dir: PathBuf, // ~
    pub store: Store,
    pub test_only: bool,
    pub working_dir: PathBuf,

    file_manager: Arc<OnceCell<ProtoFileManager>>,
    plugin_loader: Arc<OnceCell<PluginLoader>>,
}

impl ProtoEnvironment {
    pub fn new() -> Result<Self, ProtoEnvError> {
        let home = home_dir().ok_or(ProtoEnvError::MissingHomeDir)?;
        let mut root = path_var("PROTO_HOME")
            .or_else(|| path_var("XDG_DATA_HOME").map(|xdg| xdg.join("proto")))
            .unwrap_or_else(|| home.join(".proto"));

        if let Ok(rel_root) = root.strip_prefix("~") {
            root = home.join(rel_root);
        }

        Self::from(root, home)
    }

    pub fn new_testing(sandbox: &Path) -> Result<Self, ProtoEnvError> {
        let mut env = Self::from(sandbox.join(".proto"), sandbox.join(".home"))?;
        env.test_only = true;

        Ok(env)
    }

    pub fn from<R: AsRef<Path>, H: AsRef<Path>>(root: R, home: H) -> Result<Self, ProtoEnvError> {
        let root = root.as_ref();
        let home = home.as_ref();

        debug!(
            store = ?root,
            home = ?home,
            "Creating proto environment, detecting store",
        );

        Ok(ProtoEnvironment {
            config_mode: ConfigMode::Upwards,
            working_dir: env::current_dir().map_err(|_| ProtoEnvError::MissingWorkingDir)?,
            env_mode: env::var("PROTO_ENV").ok(),
            home_dir: home.to_owned(),
            file_manager: Arc::new(OnceCell::new()),
            plugin_loader: Arc::new(OnceCell::new()),
            test_only: env::var("PROTO_TEST").is_ok(),
            store: Store::new(root),
        })
    }

    pub fn get_config_dir(&self, pin: PinLocation) -> &Path {
        match pin {
            PinLocation::Global => &self.store.dir,
            PinLocation::Local => &self.working_dir,
            PinLocation::User => &self.home_dir,
        }
    }

    pub fn get_plugin_loader(&self) -> Result<&PluginLoader, ProtoConfigError> {
        let config = self.load_config()?;

        self.plugin_loader.get_or_try_init(|| {
            let mut options = config.settings.http.clone();
            options.cache_dir = Some(self.store.cache_dir.join("requests"));

            let mut loader =
                PluginLoader::new(&self.store.plugins_dir, self.store.temp_dir.join("plugins"));

            if let Some(registries) = config.settings.registries.clone() {
                loader.add_registries(registries);
            };

            if let Some(secs) = config.settings.cache_duration {
                loader.set_cache_duration(Duration::from_secs(secs));
            }

            loader.set_client_options(&options);
            loader.set_offline_checker(is_offline);

            Ok(loader)
        })
    }

    pub fn get_virtual_paths(&self) -> BTreeMap<PathBuf, PathBuf> {
        BTreeMap::from_iter([
            (self.working_dir.clone(), "/cwd".into()),
            (self.store.dir.clone(), "/proto".into()),
            (self.home_dir.clone(), "/userhome".into()),
        ])
    }

    pub fn get_virtual_paths_compat(&self) -> BTreeMap<String, PathBuf> {
        self.get_virtual_paths()
            .into_iter()
            .map(|(key, value)| (key.to_string_lossy().to_string(), value))
            .collect()
    }

    pub fn load_config(&self) -> Result<&ProtoConfig, ProtoConfigError> {
        self.load_config_with_mode(self.config_mode)
    }

    pub fn load_config_with_mode(
        &self,
        mode: ConfigMode,
    ) -> Result<&ProtoConfig, ProtoConfigError> {
        let manager = self.load_file_manager()?;

        match mode {
            ConfigMode::Global => manager.get_global_config(),
            ConfigMode::Local => manager.get_local_config(&self.working_dir),
            ConfigMode::Upwards => manager.get_merged_config_without_global(),
            ConfigMode::UpwardsGlobal => manager.get_merged_config(),
        }
    }

    pub fn load_config_files(&self) -> Result<Vec<&ProtoConfigFile>, ProtoConfigError> {
        Ok(self
            .load_file_manager()?
            .entries
            .iter()
            .filter_map(|dir| {
                if !self.config_mode.includes_global() && dir.location == PinLocation::Global
                    || self.config_mode.only_local() && dir.path != self.working_dir
                {
                    None
                } else {
                    Some(&dir.configs)
                }
            })
            .flatten()
            .collect())
    }

    #[tracing::instrument(name = "load_all", skip_all)]
    pub fn load_file_manager(&self) -> Result<&ProtoFileManager, ProtoConfigError> {
        self.file_manager.get_or_try_init(|| {
            // Don't traverse passed the home directory,
            // but only if working directory is within it!
            let end_dir = if self.working_dir.starts_with(&self.home_dir) {
                Some(self.home_dir.as_path())
            } else {
                None
            };

            let mut manager =
                ProtoFileManager::load(&self.working_dir, end_dir, self.env_mode.as_ref())?;

            // Always load the proto home/root config last
            let path = self.store.dir.join(PROTO_CONFIG_NAME);

            manager.entries.push(ProtoDirEntry {
                path: self.store.dir.clone(),
                location: PinLocation::Global,
                configs: vec![ProtoConfigFile {
                    exists: path.exists(),
                    path,
                    config: ProtoConfig::load_from(&self.store.dir, true)?,
                }],
                lock: None,
            });

            // Remove the pinned `proto` version from global/user configs,
            // as it causes massive recursion and `proto` process chains
            manager.remove_proto_pins();

            Ok(manager)
        })
    }
}

impl AsRef<ProtoEnvironment> for ProtoEnvironment {
    fn as_ref(&self) -> &ProtoEnvironment {
        self
    }
}

impl fmt::Debug for ProtoEnvironment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("ProtoEnvironment")
            .field("config_mode", &self.config_mode)
            .field("env_mode", &self.env_mode)
            .field("home_dir", &self.home_dir)
            .field("store", &self.store)
            .field("test_only", &self.test_only)
            .field("working_dir", &self.working_dir)
            .finish()
    }
}
