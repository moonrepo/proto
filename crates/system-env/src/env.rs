use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::env::{self, consts};
use std::fmt;

/// Architecture of the host environment.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[cfg_attr(feature = "schematic", derive(schematic::Schematic))]
#[serde(rename_all = "lowercase")]
pub enum SystemArch {
    X86,
    #[serde(alias = "x86_64")]
    X64,
    Arm,
    #[serde(alias = "aarch64")]
    Arm64,
    #[serde(alias = "loongarch64")]
    LongArm64,
    M68k,
    Mips,
    Mips64,
    Powerpc,
    Powerpc64,
    Riscv64,
    S390x,
    Sparc64,
}

impl SystemArch {
    pub fn from_env() -> SystemArch {
        serde_json::from_value(Value::String(consts::ARCH.to_owned()))
            .expect("Unknown architecture!")
    }

    pub fn to_rust_arch(&self) -> String {
        match self {
            Self::X64 => "x86_64".into(),
            Self::Arm64 => "aarch64".into(),
            Self::LongArm64 => "loongarch64".into(),
            _ => self.to_string(),
        }
    }
}

impl Default for SystemArch {
    #[cfg(target_arch = "wasm32")]
    fn default() -> Self {
        SystemArch::X64
    }

    #[cfg(not(target_arch = "wasm32"))]
    fn default() -> Self {
        SystemArch::from_env()
    }
}

impl fmt::Display for SystemArch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", format!("{:?}", self).to_lowercase())
    }
}

/// Operating system of the host environment.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[cfg_attr(feature = "schematic", derive(schematic::Schematic))]
#[serde(rename_all = "lowercase")]
pub enum SystemOS {
    Android,
    Dragonfly,
    FreeBSD,
    IOS,
    Linux,
    #[serde(alias = "mac")]
    MacOS,
    NetBSD,
    OpenBSD,
    Solaris,
    Windows,
}

impl SystemOS {
    pub fn from_env() -> SystemOS {
        serde_json::from_value(Value::String(consts::OS.to_owned()))
            .expect("Unknown operating system!")
    }

    pub fn is_bsd(&self) -> bool {
        matches!(
            self,
            Self::Dragonfly | Self::FreeBSD | Self::NetBSD | Self::OpenBSD
        )
    }

    pub fn is_linux(&self) -> bool {
        matches!(self, Self::Linux)
    }

    pub fn is_mac(&self) -> bool {
        matches!(self, Self::MacOS)
    }

    pub fn is_unix(&self) -> bool {
        self.is_bsd() || matches!(self, Self::Linux | Self::MacOS)
    }

    pub fn is_windows(&self) -> bool {
        matches!(self, Self::Windows)
    }

    pub fn to_rust_os(&self) -> String {
        self.to_string()
    }
}

impl Default for SystemOS {
    #[cfg(target_arch = "wasm32")]
    fn default() -> Self {
        SystemOS::Linux
    }

    #[cfg(not(target_arch = "wasm32"))]
    fn default() -> Self {
        SystemOS::from_env()
    }
}

impl fmt::Display for SystemOS {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", format!("{:?}", self).to_lowercase())
    }
}

#[cfg(windows)]
pub fn is_command_on_path(name: &str) -> bool {
    let Ok(system_path) = env::var("PATH") else {
        return false;
    };
    let Ok(path_ext) = env::var("PATHEXT") else {
        return false;
    };
    let exts = path_ext.split(';').collect::<Vec<_>>();

    for path_dir in env::split_paths(&system_path) {
        for ext in &exts {
            if path_dir.join(format!("{name}{ext}")).exists() {
                return true;
            }
        }
    }

    false
}

#[cfg(not(windows))]
pub fn is_command_on_path(name: &str) -> bool {
    let Ok(system_path) = env::var("PATH") else {
        return false;
    };

    for path_dir in env::split_paths(&system_path) {
        if path_dir.join(name).exists() {
            return true;
        }
    }

    false
}
