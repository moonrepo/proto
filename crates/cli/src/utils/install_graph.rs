use super::tool_record::ToolRecord;
use proto_core::Id;
use rustc_hash::{FxHashMap, FxHashSet};
use std::sync::Arc;
use tokio::sync::RwLock;

pub enum InstallStatus {
    ReqFailed(Id),
    WaitingOnReqs(Vec<Id>),
}

#[derive(Clone)]
pub struct InstallGraph {
    /// Tools that must be installed
    ids: Arc<FxHashSet<Id>>,

    /// Tools that have been installed
    installed: Arc<RwLock<FxHashSet<Id>>>,

    /// Tools that have been not installed (some error occurred)
    not_installed: Arc<RwLock<FxHashSet<Id>>>,

    /// Tools that require other tools to be installed
    /// before they can be installed
    requires: Arc<FxHashMap<Id, Vec<Id>>>,
}

impl InstallGraph {
    pub fn new(tools: &[ToolRecord]) -> Self {
        let mut ids = FxHashSet::default();
        let mut requires = FxHashMap::default();

        for tool in tools {
            ids.insert(tool.id.clone());

            if !tool.metadata.requires.is_empty() {
                requires.insert(
                    tool.id.clone(),
                    tool.metadata.requires.iter().map(Id::raw).collect(),
                );
            }
        }

        Self {
            ids: Arc::new(ids),
            installed: Arc::new(RwLock::new(FxHashSet::default())),
            not_installed: Arc::new(RwLock::new(FxHashSet::default())),
            requires: Arc::new(requires),
        }
    }

    pub async fn check_install_status(&self, id: &Id) -> Option<InstallStatus> {
        if !self.ids.contains(id) {
            return None;
        }

        if let Some(require_ids) = self.requires.get(id) {
            let installed = self.installed.read().await;
            let not_installed = self.not_installed.read().await;
            let mut waiting_on = vec![];

            for require_id in require_ids {
                if !self.ids.contains(require_id) {
                    continue;
                }

                if !installed.contains(require_id) {
                    waiting_on.push(require_id.clone());
                } else if not_installed.contains(require_id) {
                    return Some(InstallStatus::ReqFailed(require_id.clone()));
                }
            }

            if !waiting_on.is_empty() {
                return Some(InstallStatus::WaitingOnReqs(waiting_on));
            }
        }

        None
    }

    pub async fn mark_installed(&self, id: &Id) {
        self.installed.write().await.insert(id.to_owned());
    }

    pub async fn mark_not_installed(&self, id: &Id) {
        self.not_installed.write().await.insert(id.to_owned());
    }
}
