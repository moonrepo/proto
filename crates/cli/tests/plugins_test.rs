mod utils;

use proto_core::{
    load_tool_from_locator, Id, PluginLocator, ProtoEnvironment, Tool, UnresolvedVersionSpec,
};
use starbase_sandbox::assert_snapshot;
use std::env;
use std::fs;
use std::future::Future;
use std::path::{Path, PathBuf};
use utils::*;

async fn run_tests<F, Fut>(factory: F)
where
    F: FnOnce(&Path) -> Fut,
    Fut: Future<Output = miette::Result<Tool>>,
{
    let fixture = create_empty_sandbox();
    let proto = ProtoEnvironment::from(fixture.path()).unwrap();

    let mut tool = factory(fixture.path()).await.unwrap();

    env::set_var("PROTO_HOME", fixture.path().to_string_lossy().to_string());

    tool.setup(&UnresolvedVersionSpec::parse("1.0.0").unwrap(), false)
        .await
        .unwrap();

    env::remove_var("PROTO_HOME");

    assert!(tool.get_product_dir().exists());

    let base_dir = proto.store.inventory_dir.join("moon/1.0.0");

    if cfg!(windows) {
        assert_eq!(tool.get_exe_path().unwrap(), &base_dir.join("moon.exe"));
        assert!(proto.store.shims_dir.join("moon.exe").exists());
    } else {
        assert_eq!(tool.get_exe_path().unwrap(), &base_dir.join("moon"));
        assert!(proto.store.shims_dir.join("moon").exists());
    }
}

mod plugins {
    use super::*;

    #[tokio::test]
    async fn downloads_and_installs_plugin_from_file() {
        run_tests(|root| {
            let root_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));

            load_tool_from_locator(
                Id::raw("moon"),
                ProtoEnvironment::from(root).unwrap(),
                PluginLocator::SourceFile {
                    file: "./tests/fixtures/moon-schema.toml".into(),
                    path: root_dir.join("./tests/fixtures/moon-schema.toml"),
                },
            )
        })
        .await;
    }

    #[tokio::test]
    #[should_panic(expected = "does not exist")]
    async fn errors_for_missing_file() {
        run_tests(|root| {
            let root_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));

            load_tool_from_locator(
                Id::raw("moon"),
                ProtoEnvironment::from(root).unwrap(),
                PluginLocator::SourceFile {
                    file: "./some/fake/path.toml".into(),
                    path: root_dir.join("./some/fake/path.toml"),
                },
            )
        })
        .await;
    }

    #[tokio::test]
    async fn downloads_and_installs_plugin_from_url() {
        run_tests(|root| {
            load_tool_from_locator(
                Id::raw("moon"),
                ProtoEnvironment::from(root).unwrap(),
                PluginLocator::SourceUrl {
                    url: "https://raw.githubusercontent.com/moonrepo/moon/master/proto-plugin.toml"
                        .into(),
                },
            )
        })
        .await;
    }

    #[tokio::test]
    #[should_panic(expected = "does not exist")]
    async fn errors_for_broken_url() {
        run_tests(|root| {
            load_tool_from_locator(
                Id::raw("moon"),
                ProtoEnvironment::from(root).unwrap(),
                PluginLocator::SourceUrl {
                    url: "https://raw.githubusercontent.com/moonrepo/moon/some/fake/path.toml"
                        .into(),
                },
            )
        })
        .await;
    }

    mod builtins {
        use super::*;

        // macos is very flaky!
        #[cfg(not(target_os = "macos"))]
        #[test]
        fn supports_bun() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("bun")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "bun")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_deno() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("deno")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "deno")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_go() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("go")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "go")
                .arg("version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_node() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("node")
                .arg("--")
                .arg("--no-bundled-npm")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "node")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_npm() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("node")
                .arg("--")
                .arg("--no-bundled-npm")
                .assert()
                .success();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("npm")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "npm")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_pnpm() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("node")
                .arg("--")
                .arg("--no-bundled-npm")
                .assert()
                .success();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("pnpm")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "pnpm")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_yarn() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("node")
                .arg("--")
                .arg("--no-bundled-npm")
                .assert()
                .success();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("yarn")
                .assert()
                .success();

            create_shim_command(sandbox.path(), "yarn")
                .arg("--version")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_python() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("python")
                .arg("3.12.0") // Latest doesn't always work
                .assert()
                .success();

            create_shim_command(sandbox.path(), "python")
                .arg("--version")
                .env("PROTO_PYTHON_VERSION", "3.12.0")
                .assert()
                .success();

            assert_snapshot!(
                fs::read_to_string(sandbox.path().join(".proto/shims/registry.json")).unwrap()
            );
        }

        #[test]
        fn supports_rust() {
            let sandbox = create_empty_sandbox();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("rust")
                .assert();

            // Doesn't create shims
        }

        #[test]
        fn supports_toml_schema() {
            let sandbox = create_empty_sandbox_with_tools();

            create_proto_command(sandbox.path())
                .arg("install")
                .arg("moon-test")
                .assert()
                .success();

            // Doesn't create shims
        }
    }
}
