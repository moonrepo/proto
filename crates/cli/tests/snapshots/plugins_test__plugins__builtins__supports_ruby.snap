---
source: crates/cli/tests/plugins_test.rs
expression: "fs::read_to_string(sandbox.path().join(\".proto/shims/registry.json\")).unwrap()"
---
{
  "bundle": {
    "alt_bin": true,
    "parent": "ruby"
  },
  "gem": {
    "alt_bin": true,
    "parent": "ruby"
  },
  "rake": {
    "alt_bin": true,
    "parent": "ruby"
  },
  "ruby": {}
}
